### ISSUE TITLE
3.5.0: Various rendering issues related to tint, fades and flips.

### ISSUE THREAD

**Issue by sonneveld** [adventuregamestudio/ags#865]
Some rendering issues found:

 -  If screen tint is enabled, fades like dissolve and box will fade out instead and then not fade back in, effectively making game unplayable.
 -  if `SetFadeColour` is set with values ==255, like (255,0,255) , then it will not fade out, instead you get a pause then a blank screen.  254 and other values are fine though (edit: I realised this is the mask colour, so I don't know what the correct behavior should be)
 - fading out to a colour will work, but then will leave the screen black instead of the required colour. 
 - With software/8bit, the fade out will not show a black screen, but instead the normal screen.
 - flipping screen will show black areas if character and room is scrolled to right side of right.
 - software driver driver crashes with cross fade screen transition.

I've tested with AGS 3.5.0-beta5 on Windows. Issues can be found in all the drivers but I haven't tested all combinations.

Attached is a demonstration game that lets you change screen and character rendering options, along with a wide room to test scrolling.
[screentest-v3-20190706.zip](https://github.com/adventuregamestudio/ags/files/3364368/screentest-v3-20190706.zip)
Source included so you can change to 8, 16 and 32bit.

**Comment by sonneveld** [comment]
There's a lot of combinations to tests when you include software driver and colour depth so it might be worth being systematic and drawing up a test table for these things. I'll wait and see if we find any other issues to include.

**Comment by sonneveld** [comment]
Changing the width of the camera or viewport after it is displayed seems to create a corrupted display with the software graphics driver.
![image](https://user-images.githubusercontent.com/335451/60753536-3ecfab80-a017-11e9-8392-f9b40355bb2e.png)

**Comment by sonneveld** [comment]
Latest test game with camera controls: 
[screentest-v4-20190706.zip](https://github.com/adventuregamestudio/ags/files/3364534/screentest-v4-20190706.zip)

**Comment by ivan-mogilko** [comment]
Could you move camera problems to separate ticket please?
Also, please elaborate what this means : `scrolling background will overwrite camera`

**Comment by sonneveld** [comment]
> Could you move camera problems to separate ticket please?

Done!

**Comment by sonneveld** [comment]
In regards to the fade out not clearing the screen at the end, looks like the original engine cleared the draw list if faded out: 

```
void render_to_screen(BITMAP *toRender, int atx, int aty) {

  atx += get_screen_x_adjustment(toRender);
  aty += get_screen_y_adjustment(toRender);
  gfxDriver->SetRenderOffset(atx, aty);

  render_black_borders(atx, aty);

  gfxDriver->DrawSprite(AGSE_FINALSCREENDRAW, 0, NULL);

  if (play.screen_is_faded_out)
  {
    if (gfxDriver->UsesMemoryBackBuffer())
      gfxDriver->RenderToBackBuffer();
    gfxDriver->ClearDrawList();
    return;
  }

```

https://github.com/sonneveld/ags-cj/blob/master/Engine/ac.cpp#L2721

**Comment by sonneveld** [comment]
So I think I know why screen tint affects fading/transitions. There's the obvious palette wrangling for 8bit mode. The other reason is that dissolve and crossfade takes a screenshot (instead of fading out), and then is used for fade in. If the screenshot contains the tint, then it would mean the tint is applied twice during the fade in, once in the screenshot and twice during the normal render of fade in.

**Comment by ivan-mogilko** [comment]
Made a quick test of all standard transitions (without tint for now).
All seem work, except for BlackBoxOut in _software mode_ which skips box-in stage.

> software driver driver crashes with cross fade screen transition

Does not seem to reproduce with the latest release-3.5.0

**Comment by sonneveld** [comment]
> Does not seem to reproduce with the latest release-3.5.0

Are you suggesting these issues were fixed with the latest changes? Or that you were unable to reproduce?

**Comment by ivan-mogilko** [comment]
> Are you suggesting these issues were fixed with the latest changes? Or that you were unable to reproduce?

I tried with 3.5.0.14 tag and still cannot reproduce.

**Comment by ivan-mogilko** [comment]
I pushed a fix for BlackBoxOut in software mode.

Investigating the code, it appears that fade transition is explicitly disabled if screen tint is active.

**Comment by ivan-mogilko** [comment]
> We would still need to render a coloured quad on top of or instead of the normal constructed screen.

I think it might be possible to reuse `_screenTintSprite` texture from D3D and OGL drivers.

**Comment by ivan-mogilko** [comment]
Alright, I reimplemented Tint and Fade (complete overlay) as a sprite entry in all drivers (software one is a little hacky).

Also made sure that -
* Fade always covers the Tint effect (there was some discrepancy between drivers)
* Neither fade nor tint actually cover the engine's overlay (FPS, console)

**Comment by ivan-mogilko** [comment]
@sonneveld , I could not reproduce this either:
> flipping screen will show black areas if character and room is scrolled to right side of right.

Maybe I am doing it wrong, in which case I need more details.

I believe that I either fixed or was not able to reproduce all of the issues mentioned in the first post. Could you make a test when you have a spare time?

**Comment by sonneveld** [comment]
These are my notes from testing `75b1687` against the screentest app:

```
common:
- NEW bug: during black fade out, noticed some screen corruption instead of black screen. (in all drivers)
- NEW bug: moving straight up while in middle of scrolled room will result in the character returning in an unexpected position (top left instead of middle)
- screen tint prevents any transition effect.

in software
- box in and out animations differ in shape, they should be same ratio as screen
- FIXED: could not reproduce black sides when flipping
- FIXED: no cross fade transition crashing.

in direct3d/opengl:
- fade out borked mode does not work, just pauses but screen remains there.
- BUG: visible black sides when flipping. To replicate, move all the way to the right side of the screentest room and use the FlipScreen button. Actually, it is replicatable without moving the character, which is initially positioned in the middle of scrolled room.
```

**Comment by ivan-mogilko** [comment]
> screen tint prevents any transition effect.

Like I mentioned above, this is how it is deliberatly coded right now.

**Comment by sonneveld** [comment]
This is an example of the fade in/out corruption I see occasionally with fade out to black:

![image](https://user-images.githubusercontent.com/335451/62291897-6ac03e80-b4a8-11e9-9a16-1c7db5c25241.png)

**Comment by sonneveld** [comment]
> PS. Also, FlipScreen is one of those weird features which cause headache to player, because they also flip cursor movement. Have you ever seen a game that uses this? I cannot remember single one.

I can try to hunt for a game in the game archives. You're right, I can only see it being useful in a cutscene unless the mouse cursor was rendered on top of the flipped screen in earlier engines.

**Comment by ivan-mogilko** [comment]
I pushed a fix (hopefully) for fadeout error: from what I found, the created quad was not actually painted in case of black colour (operation skipped).

**Comment by ivan-mogilko** [comment]
> NEW bug: moving straight up while in middle of scrolled room will result in the character returning in an unexpected position (top left instead of middle)

Is this a display issue? Maybe something's wrong with Character.ChangeRoomAutoPosition command? Need to investigate how it supposed to work.

**Comment by sonneveld** [comment]
> Is this a display issue? Maybe something's wrong with Character.ChangeRoomAutoPosition command? Need to investigate how it supposed to work.

It might be related to the camera fixes actually. I did not notice this issue with the previous builds.

**Comment by ivan-mogilko** [comment]
> box in and out animations differ in shape, they should be same ratio as screen

I honestly don't see anything, are there any screenshots that illustrate this?

**Comment by ivan-mogilko** [comment]
> fade out borked mode does not work, just pauses but screen remains there.

What is this "borked mode"? I checked the script, and it looks like you are setting transparent fade colour. How was this intended to look like?

**Comment by ivan-mogilko** [comment]
> fade out borked mode does not work, just pauses but screen remains there.

This is due the rules of conversion from bitmap to texture, 255,0,255 is converted to either black colour on opaque surfaces or fully transparent colour (alpha = 0) on non-opaque ones. I'd rather ask why does it show anything with software renderer (probably because it uses special blit function which does not support transparency).

**Comment by sonneveld** [comment]
> I honestly don't see anything, are there any screenshots that illustrate this?

aw I just recorded videos for this :) software: https://www.youtube.com/watch?v=4OBWs2Q1kgs vs opengl: https://www.youtube.com/watch?v=Ga8kwW3CoRk

> This is due the rules of conversion from bitmap to texture 

Should these rules still apply for textures that did not originate from an allegro bitmap?

**Comment by ivan-mogilko** [comment]
> maybe in the texture conversion, instead of opaque true/false to determine the replacement colour, the argument could be an RGB replacement value.

This idea on its own may have its value, but at the moment the only case seem to be the fade colour, everywhere else it's considered transparent colour. I am not sure it's conceptually correct, neither that it's worth implementing special case (right now or ever). IMO question also is, how should SetFadeColour react if passed what is considered a transparent colour by the rest of functions.

**Comment by ivan-mogilko** [comment]
> maybe in the texture conversion, instead of opaque true/false to determine the replacement colour, the argument could be an RGB replacement value.

So what I was going to try was slightly different solution: just add a flag that tells to ignore mask color. But as I started working on this, it suddenly hit me, is not that it's what "opaque" flag means in the first place?

If you search for "opaque" usage, you may notice that software renderer uses "copy" blit method for opaque sprites, which means that mask color is ignored completely.

This means that accelerated renderers actually had a mistake in their conversion to texture. Instead of filling mask pixels with black in opaque mode they should've just ignore mask color.

**Comment by ivan-mogilko** [comment]
So, I was investigating this further, and apparently accelerated renderers are doing it wrong with "opaque" flag. Judging by how software mode works, opaque means that mask color is completely ignored and colors blit as-is. Meanwhile D3D and OpenGL are replacing mask color with black.

This discrepancy is easily seen if you draw 255,0,255 shape on a room background, which is drawn as "opaque": software renderer retains pink color while D3D and OGL will have it black.
So this is a bug on its own.

There's also a slightly confusing logic originating in software mode, which sais that opaque sprites cannot be drawn translucent (extra transparency option). At the same time, D3D and OGL are happy to render "opaque" sprites with transparency too without problem (they simply do not check for this flag anymore after converting bitmap to texture).

I think originally this "opaque" sprite mode was meant for something that cannot be transparent or translucent ever, like room backgrounds or special bitmaps (existing examples are: black borders, debug mask overlay, debug console, video playback frame).

My current thinking is to actually allow translucent drawing of opaque sprites in software mode (even though it's not used in practice atm). Then, fix opaque sprites conversion in accelerated mode.
Then - set fade and screen tint sprites as "opaque" universally (they are marked as non-opaque atm, perhaps to tell that they are allowed to be translucent).

In other words, after this change "opaque" flag would mean strictly "ignore mask color", while it may be still drawn translucent if transparency parameter is set.

**Comment by sonneveld** [comment]
Would this affect 32bit sprites with alpha channels? We modify sprites on load, replacing transparent pixels with the mask colour. I don't think they would ever get rendered as "opaque" though.

**Comment by ivan-mogilko** [comment]
> Would this affect 32bit sprites with alpha channels? We modify sprites on load, replacing transparent pixels with the mask colour. I don't think they would ever get rendered as "opaque" though.

This should not affect them in any way as my point is to *stop changing mask colour to black* in sprite marked as "opaque".

**Comment by ivan-mogilko** [comment]
I pushed couple of commits, the changes in a nutshell are:
* accelerated renderers treat fade and tint effects as opaque sprites;
* optimised function for converting opaque bitmap into texture (skipping all masked color checks).

**Comment by ivan-mogilko** [comment]
Regarding the screen flip, I think I found the fix (merging flip operation with sprite batch transform), but there's another thing that has to be fixed first: the order of drawing is still slightly wrong, game cursor is now drawn above debugging overlay (fps counter & console), and screen flip requires that game cursor's position is flipped too, but debug overlay probably should not be flipped (also I just found that I may have broken plugins a little by drawing fps before AGSE_POSTSCREENDRAW plugin hook).

Things are somewhat complicated by the fact that drawing is split into 2 parts, one of which is not performed in certain circumstances (such as fade-ins, and Display command), so it has to be carefuly untangled again.

**Comment by sonneveld** [comment]
Another option might be to drop software rendering altogether. Or just render everything to a backing bitmap, that gets flipped before being blitted to the screen? That would mean the dirty rectangle system wouldn't be necessary.

Or we allow the debug overlay to be flipped? I'm not sure how much people use the overlay. And if flipping is only meant to be used for cutscenes (or possibly only with keyboard inputs?), maybe cutscenes aren't debugged?

**Comment by ivan-mogilko** [comment]
> Another option might be to drop software rendering altogether.

This is obviously not an option for this release.

There's too much going on here, and dirty rects also get into way, because normally they are optimized to be stored right on the virtual screen... which means we cannot flip virtual screen itself.
This could be solved probably with certain effort, but I am leaning towards ignoring this issue and just let software renderer flip everything unconditionally. Becuase I don't know how much drawing will change if/when ags will use SDL backend (does it use accelerated 2D graphics?) and maybe all these optimizations become unnecessary and issue resolve on its own.

Anyway, while researching the flip problem I found there's more incorrect bits left behind after viewport/camera update, so currently I am trying to tidy drawing order once more.

And, for example, ShakeScreen is still implemented using "global sprite offset" which will give incorrect results with certain transforms.

**Comment by ivan-mogilko** [comment]
An update: I pushed several commits which include:
* D3D and OGL support to have screen flip and global offset (used by screen shake) only for distinct sprite groups.;
* Tidied order of elements once more, and also made debug overlay to be drawn above any game effects (tint, fade, shake etc).
The latter was not particulary necessary, but I think it makes it more clear to understand what's being drawn.

There are following two known issues remaining, *which are old issues present in previous versions*:
* software renderer cannot have flip and fade effect at the same time (flip gets disabled for the fade out time);
* software renderer cannot have distinct sprite layers (like fps counter) which are not affected by screen flip;

The first could be fixed relatively easy (but also may be worked around using scripted effects), the second will require introducing tree-like scene node structure and more intermediate buffers when composing virtual screen in software mode. I guess it's all doable, but after couple attempts I realized that don't want to risk screwing things up even more so close to release.

I was also very much willing to move FadeIn/Out and BlackBoxOut code out of the graphic drivers and make just a collection of clear functions for these screen effects, but then remembered the rule about not refactoring much unless needed for fix, so restrained myself.

A bit later I shall perhaps write couple of tickets with my thoughts on how the rendering logic could be cleaned further.

**Comment by i30817** [comment]
Is this the same issue that affects this old game (well demo): https://archive.org/details/super-jazz-man-demo

To be clear if you use the hardware driver in opengl most of the background is missing/not painted. Software 'fixes' it at the cost of being slow.

**Comment by ivan-mogilko** [comment]
> Is this the same issue that affects this old game? (well demo, it's just 1mb, easy to try): https://archive.org/details/super-jazz-man-demo
> 
> To be clear if you use the hardware driver in opengl most of the background is missing/not painted. Software 'fixes' it at the cost of being slow.

Only quickly checked this out, and it looks like something different, more like a z-order issue. I will try again bit later.

**Comment by ivan-mogilko** [comment]
Anyhow, I will close this ticket, as it's been a while and I believe most if not all of the stuff mentioned above was fixed. If there are other similar problems they may be adressed separately.
